File: go/internal/e2e/subagent_hooks_e2e_test.go
package e2e

import (
	"os"
	"path/filepath"
	"sync"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestSubagentHooksE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient(func(o *copilot.ClientOptions) {
		o.Env = append(o.Env, "COPILOT_EXP_COPILOT_CLI_SESSION_BASED_SUBAGENTS=true")
	})
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should invoke preToolUse and postToolUse hooks for sub-agent tool calls", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		type hookEntry struct {
			kind      string
			toolName  string
			sessionID string
		}
		var hookLog []hookEntry
		var mu sync.Mutex

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Hooks: &copilot.SessionHooks{
				OnPreToolUse: func(input copilot.PreToolUseHookInput, invocation copilot.HookInvocation) (*copilot.PreToolUseHookOutput, error) {
					mu.Lock()
					hookLog = append(hookLog, hookEntry{kind: "pre", toolName: input.ToolName, sessionID: input.SessionID})
					mu.Unlock()
					return &copilot.PreToolUseHookOutput{PermissionDecision: "allow"}, nil
				},
				OnPostToolUse: func(input copilot.PostToolUseHookInput, invocation copilot.HookInvocation) (*copilot.PostToolUseHookOutput, error) {
					mu.Lock()
					hookLog = append(hookLog, hookEntry{kind: "post", toolName: input.ToolName, sessionID: input.SessionID})
					mu.Unlock()
					return nil, nil
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Create a file for the sub-agent to read
		testFile := filepath.Join(ctx.WorkDir, "subagent-test.txt")
		if err := os.WriteFile(testFile, []byte("Hello from subagent test!"), 0644); err != nil {
			t.Fatalf("Failed to write test file: %v", err)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the task tool to spawn an explore agent that reads the file subagent-test.txt in the current directory and reports its contents. You must use the task tool.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		mu.Lock()
		defer mu.Unlock()

		// Parent tool hooks fire for "task"
		var taskPre *hookEntry
		for i := range hookLog {
			if hookLog[i].kind == "pre" && hookLog[i].toolName == "task" {
				taskPre = &hookLog[i]
				break
			}
		}
		if taskPre == nil {
			t.Fatal("preToolUse should fire for the parent's 'task' tool call")
		}

		// Sub-agent tool hooks fire for "view"
		var viewPre, viewPost []hookEntry
		for _, h := range hookLog {
			if h.toolName == "view" {
				if h.kind == "pre" {
					viewPre = append(viewPre, h)
				} else {
					viewPost = append(viewPost, h)
				}
			}
		}
		if len(viewPre) == 0 {
			t.Fatal("preToolUse should fire for the sub-agent's 'view' tool call")
		}
		if len(viewPost) == 0 {
			t.Fatal("postToolUse should fire for the sub-agent's 'view' tool call")
		}

		// input.SessionID distinguishes parent from sub-agent
		if viewPre[0].sessionID == taskPre.sessionID {
			t.Error("Sub-agent tool hooks should have a different sessionId than parent tool hooks")
		}
	})
}

File: nodejs/test/e2e/subagent_hooks.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { writeFile } from "fs/promises";
import { join } from "path";
import { describe, expect, it } from "vitest";
import type {
    PreToolUseHookInput,
    PreToolUseHookOutput,
    PostToolUseHookInput,
    PostToolUseHookOutput,
} from "../../src/index.js";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext, isCI } from "./harness/sdkTestContext.js";

describe("Subagent hooks", async () => {
    // For snapshot recording (non-CI), use RECORD_GH_TOKEN if available
    const recordToken = !isCI ? process.env.RECORD_GH_TOKEN : undefined;
    const {
        copilotClient: client,
        workDir,
        env,
    } = await createSdkTestContext({
        ...(recordToken ? { copilotClientOptions: { gitHubToken: recordToken } } : {}),
    });
    // Sub-agent hook propagation requires the session-based subagents feature flag.
    // Without this flag, the legacy callback-bridge path is used, which does not
    // support SDK preToolUse/postToolUse hooks for sub-agent tool calls.
    env.COPILOT_EXP_COPILOT_CLI_SESSION_BASED_SUBAGENTS = "true";

    it("should invoke preToolUse and postToolUse hooks for sub-agent tool calls", async () => {
        const hookLog: { kind: "pre" | "post"; toolName: string; sessionId: string }[] = [];

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            hooks: {
                onPreToolUse: async (input: PreToolUseHookInput) => {
                    hookLog.push({
                        kind: "pre",
                        toolName: input.toolName,
                        sessionId: input.sessionId,
                    });
                    return { permissionDecision: "allow" } as PreToolUseHookOutput;
                },
                onPostToolUse: async (input: PostToolUseHookInput) => {
                    hookLog.push({
                        kind: "post",
                        toolName: input.toolName,
                        sessionId: input.sessionId,
                    });
                    return null as PostToolUseHookOutput;
                },
            },
        });

        // Create a file for the sub-agent to read
        await writeFile(join(workDir, "subagent-test.txt"), "Hello from subagent test!");

        await session.sendAndWait({
            prompt: "Use the task tool to spawn an explore agent that reads the file subagent-test.txt in the current directory and reports its contents. You must use the task tool.",
        });

        // Parent tool hooks fire for "task"
        const taskPre = hookLog.find((h) => h.kind === "pre" && h.toolName === "task");
        expect(taskPre, "preToolUse should fire for the parent's 'task' tool call").toBeDefined();

        // Sub-agent tool hooks fire for "view"
        const viewPre = hookLog.filter((h) => h.kind === "pre" && h.toolName === "view");
        const viewPost = hookLog.filter((h) => h.kind === "post" && h.toolName === "view");
        expect(
            viewPre.length,
            "preToolUse should fire for the sub-agent's 'view' tool call"
        ).toBeGreaterThan(0);
        expect(
            viewPost.length,
            "postToolUse should fire for the sub-agent's 'view' tool call"
        ).toBeGreaterThan(0);

        // input.sessionId distinguishes parent from sub-agent: parent tools and
        // sub-agent tools carry different sessionIds
        expect(viewPre[0].sessionId).not.toBe(taskPre!.sessionId);

        await session.disconnect();
    }, 120_000);
});
